deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -147,80 +134,54 @@ def upload_screenshots(screenshots_per_language, localizations, options)
+        checksum = UploadScreenshots.calculate_checksum(screenshot.path)
+        duplicate = app_screenshot_set.app_screenshots.any? { |s| s.source_file_checksum == checksum }
 
-        # Create map to find screenshot set to add screenshot to
-        app_screenshot_sets_map = {}
-        app_screenshot_sets = localization.get_app_screenshot_sets
-        app_screenshot_sets.each do |app_screenshot_set|
-          app_screenshot_sets_map[app_screenshot_set.screenshot_display_type] = app_screenshot_set
-
-          # Set initial screnshot count
-          indized[localization.locale][app_screenshot_set.screenshot_display_type] ||= {
-            count: app_screenshot_set.app_screenshots.size,
-            checksums: []
-          }
-
-          checksums = app_screenshot_set.app_screenshots.map(&:source_file_checksum).uniq
-          indized[localization.locale][app_screenshot_set.screenshot_display_type][:checksums] = checksums
+        # Enqueue uploading job if it's not duplicated otherwise screenshot will be skipped
+        if duplicate
+          UI.message("Previous uploaded. Skipping '#{screenshot.path}'...")
+        else
+          worker.enqueue(UploadScreenshotJob.new(app_screenshot_set, screenshot.path, wait_for_processing))
         end
+      end
 
-        UI.message("Uploading #{screenshots_for_language.length} screenshots for language #{language}")
-        screenshots_for_language.each do |screenshot|
-          display_type = screenshot.device_type
-          set = app_screenshot_sets_map[display_type]
-
-          if display_type.nil?
-            UI.error("Error... Screenshot size #{screenshot.screen_size} not valid for App Store Connect")
-            next
-          end
-
-          unless set
-            set = localization.create_app_screenshot_set(attributes: {
-              screenshotDisplayType: display_type
-            })
-            app_screenshot_sets_map[display_type] = set
-
-            indized[localization.locale][set.screenshot_display_type] = {
-              count: 0,
-              checksums: []
-            }
-          end
-
-          index = indized[localization.locale][set.screenshot_display_type][:count]
+      worker.start
+    end
 
-          if index >= 10
-            UI.error("Too many screenshots found for device '#{screenshot.device_type}' in '#{screenshot.language}', skipping this one (#{screenshot.path})")
-            next
-          end
+    def clean_up_screenshots(localizations)
+      iterator = AppScreenshotIterator.new(localizations)
 
-          bytes = File.binread(screenshot.path)
-          checksum = Digest::MD5.hexdigest(bytes)
-          duplicate = indized[localization.locale][set.screenshot_display_type][:checksums].include?(checksum)
+      # Delete bad entries that are left as placeholder for some reasons, for example
+      iterator.each_app_screenshot do |_, _, app_screenshot|
+        app_screenshot.delete! unless app_screenshot.complete?
+      end
```

Comment:
I think we either need to force `wait_for_processing` to always `true` or skip this delete cleanup if its set to `false`

Reply:
> This may result in screenshots getting deleted if wait_for_processing is set to false here 👉 

Great point👍 I actually haven't considered about `wait_for_processing` that much and tested always with `false`. So I don't know how much it will slow down yet. Let me briefly test it with `wait_for_processing = true`.
There is a little chance that this random `sleep` helped me to not observe accidental deletions while I was testing 🙈  

https://github.com/fastlane/fastlane/pull/16972/files#diff-1d448c7e200ec0facd4ffd5c23cd3fdeR64

Alternatively, I can add a loop to wait until the number of completed screenshots matches the number of listed screenshots including skipped and enqueued ones by the iterator. It's going to be something like you added to `delete_screenshots` to verify the count. If it works, then performance is likely to be better than forcing `wait_for_processing =  true` since 
the number of network requests for `get_app_screenshot_sets` would be fewer than calling ` Spaceship::ConnectAPI.get_app_screenshot` individually.

Reply:
Sounds good! Really appreciate you looking into 😊 

1. Always waiting is _probably_ the safest option
2. Could optionally delete based upon what `wait_for_processing` is set to
3. We could add more statuses to https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app_screenshot.rb#L39-L49 if we wanted to check something other than `completed?`

Just ping me here or DM me on Twitter if you want to discuss it 💪

Reply:
@joshdholtz 

I came up with a better idea for it.
https://github.com/fastlane/fastlane/pull/16972/commits/af9f5de48c37d54aa705877d14aaa7c845b20489

1. Set `false` to `wait_for_procssesing` always (if it's enabled, it takes about 2 times longer on my project😅)
2. Add a loop to wait for until none of the screenshots remains "UPLOAD_COMPLETE" state
3. After the loop, check if there is an error or missing screenshot. If so, then retry.
4. No more removing screenshots without confident at `clean_up_screenshots` 

How does that sound to you? I'm still struggling with unit tests to cover those flow as it got tricky to go through verification part with mocks 😛 but I hope that direction is okay🤞

Reply:
> We could add more statuses to https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app_screenshot.rb#L39-L49 if we wanted to check something other than completed?

I ended up not using it in this PR but probably a good idea to add `processing?` for `UPLOAD_COMPLETE`

Reply:
@ainame I think this mostly makes sense! My one concern is still **automatically deleting screenshots that have failed** 🤔 Screenshots that have failed might be due to a bad image format provided by the user and I don't think we want to just delete those. I think we want to actually **fail the process** so the user can fix the issues. 

Users will submit screenshots and metadata, upload the app, and then submit for review. We want to make the user has exactly what they want up on the App Store at the end of this 💯 

Thoughts on this? 😇

Reply:
@joshdholtz 

> My one concern is still automatically deleting screenshots that have failed 🤔 Screenshots that have failed might be due to a bad image format provided by the user and I don't think we want to just delete those. I think we want to actually fail the process so the user can fix the issues.

Although screenshots being "FAILED" state would be automatically removed at the end of `upload_screenshots` before retrying, I introduced the retry limit which leads `UI.user_error!` eventually if it can not avoid failures. And also I expect  `upload_screenshots` with retry is going to skip most of the succeeded screenshots and only try previously failed ones so that it doesn't take up time.

To provide more information about what failed to users, I can move this removing code bit above [the retry](https://github.com/fastlane/fastlane/pull/16972/files#diff-1d448c7e200ec0facd4ffd5c23cd3fdeR182) to leave errors on App Store Connect in the end?

What do you think?

Reply:
Ohhhhhh, I totally didn't put together that things would fail after a retry 😛 What you have should be good then!